File: hydra-python/src/main/python/hydra/dsl/lib/pairs.py
"""Phantom-typed term DSL for the hydra.lib.pairs library."""

from hydra.phantoms import TTerm
from hydra.dsl.phantoms import lib_primitive1 as primitive1


def first(pair: TTerm) -> TTerm:
    """
    Get the first element of a pair.

    Type: (a, b) -> a
    """
    return primitive1(pair)


def second(pair: TTerm) -> TTerm:
    """
    Get the second element of a pair.

    Type: (a, b) -> b
    """
    return primitive1(pair)

File: hydra-haskell/src/main/haskell/Hydra/Dsl/Lib/Pairs.hs
-- | Phantom-typed term DSL for the hydra.lib.pairs library

module Hydra.Dsl.Lib.Pairs where

import Hydra.Phantoms
import Hydra.Dsl.Meta.Phantoms
import qualified Hydra.Dsl.Terms as Terms
import Hydra.Sources.Libraries

first :: TTerm (a, b) -> TTerm a
first = primitive1 _pairs_first

second :: TTerm (a, b) -> TTerm b
second = primitive1 _pairs_second
